### PR TITLE
[WIP] `IndexIterator` as template parameter of `Indexer`

### PR DESCRIPTION
**Pull request by emcastillo**
Avoids having cuda specific code in chainerx generic index iterator class.

### CODE REVIEW

**Comment by asi1024** [comment]
Could you fix for C++ lint check?

File: chainerx_cc/chainerx/indexer.h [hunk 1]
```diff
@@ -37,16 +37,16 @@ class NdimIndex {
     int8_t ndim_;
 };
 
-template <int8_t kNdim = kDynamicNdim>
+template <int8_t kNdim = kDynamicNdim, class T = IndexIterator<kNdim>>
```
Comment (asi1024) [inline chainerx_cc/chainerx/indexer.h]: I'm not confident that it is possible, but can we specify the index iterator with a member type?
```cuda
#ifdef __CUDA_ARCH__
using indexer_type = CudaIndexIterator<kNdim>;
#else
using indexer_type = IndexIterator<kNdim>;
#endif
```

**Comment by emcastillo** [comment]
Need to fix several other issues too 😂

File: chainerx_cc/chainerx/indexer.h (hunk 1 above)
Comment (emcastillo) [reply chainerx_cc/chainerx/indexer.h]: Mm this would add a dependence of the `indexer.h` class to the `cuda/index_iterator` code

**Comment by asi1024** [comment]
According to @emcastillo, the building time increased a lot because of the templating.

**Pull request closed without merging**